pootle/apps/pootle_app/management/commands/calculate_checks.py

```diff
@@ -14,69 +14,98 @@
-    # URLs need to be aware of that and they are cached. Ideally
-    # Django should take care of setting this up, but it doesn't yet:
-    # https://code.djangoproject.com/ticket/16734
-    script_name = (u'/' if settings.FORCE_SCRIPT_NAME is None
-                        else force_unicode(settings.FORCE_SCRIPT_NAME))
-    set_script_prefix(script_name)
-    super(RefreshStatsCommand, Command()).handle_noargs(**options)
+    cached_methods = [CachedMethods.CHECKS]
+    process_disabled_projects = True
+
+    def handle_all_stores(self, translation_project, **options):
+        self.process(translation_project=translation_project, **options)
+
+    def handle_all(self, **options):
+        if not self.projects and not self.languages:
+            logging.info(u"Running %s (noargs)", self.name)
+            try:
+                self.process(**options)
+            except Exception:
+                logging.exception(u"Failed to run %s", self.name)
+        else:
+            super(Command, self).handle_all(**options)
+
+    @staticmethod
+    def process(**options):
+        check_names = options.get('check_names', None)
+        translation_project = options.get('translation_project', None)
+        store_fk_filter = {}
+        unit_fk_filter = {}
+
+        if translation_project is not None:
+            store_fk_filter = {
+                'store__translation_project': translation_project,
+            }
+            unit_fk_filter = {
+                'unit__store__translation_project': translation_project,
+            }
+
+        logging.info('Calculating quality checks for all units...')
+        QualityCheck.delete_unknown_checks()
+
+        checks = QualityCheck.objects.filter(**unit_fk_filter)
+        if check_names is not None:
+            checks = checks.filter(name__in=check_names)
+        checks = checks.values('id', 'name', 'unit_id',
+                               'category', 'false_positive')
+        all_units_checks = {}
+        for check in checks:
+            all_units_checks.setdefault(check['unit_id'], {})[check['name']] = check
+
+        unit_count = 0
+        # unit's query is faster without `select_related('store')`
+        units = Unit.simple_objects.filter(state__gt=OBSOLETE) \
+                                   .order_by('store__id')
+
+        store = None
+        # units are ordered by store, we update dirty cache after we switch
+        # to another store
+        for unit in units.filter(**store_fk_filter).iterator():
+            if store is None or unit.store_id != store.id:
+                if store is not None:
+                    store.update_dirty_cache()
+                # we get unit.store only if the store differs from previous
+                store = Store.simple_objects.get(id=unit.store_id)
+
+            # HACKISH: set unit.store to avoid extra querying in
+            # `unit.update_quality_checks()` method
+            unit.store = store
+            unit_count += 1
```

Comment:
I would add a blank line before this, just to visually separate the hackish section from the regular code.